Race when deregistering MBeans

There is likely a race between MBean deregistration in shutdown and after rendering, leading to the following error:

```
Exception in thread “hz._hzInstance_2_jet.cached.thread-2” java.lang.RuntimeException: javax.management.InstanceNotFoundException: com.hazelcast:type=Metrics,instance=_hzInstance_2_jet,tag0=tcp,tag1=“connection[/192.168.0.2:49890->/192.168.0.2:5701]“,tag2=in
    at com.hazelcast.jet.impl.metrics.jmx.JmxPublisher.whenComplete(JmxPublisher.java:109)
    at com.hazelcast.jet.impl.metrics.JetMetricsService.lambda$init$0(JetMetricsService.java:98)
    at com.hazelcast.spi.impl.executionservice.impl.DelegateAndSkipOnConcurrentExecutionDecorator$DelegateDecorator.run(DelegateAndSkipOnConcurrentExecutionDecorator.java:66)
    at com.hazelcast.util.executor.CachedExecutorServiceDelegate$Worker.run(CachedExecutorServiceDelegate.java:227)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
    at java.lang.Thread.run(Thread.java:748)
    at com.hazelcast.util.executor.HazelcastManagedThread.executeRun(HazelcastManagedThread.java:64)
    at com.hazelcast.util.executor.HazelcastManagedThread.run(HazelcastManagedThread.java:80)
Caused by: javax.management.InstanceNotFoundException: com.hazelcast:type=Metrics,instance=_hzInstance_2_jet,tag0=tcp,tag1=“connection[/192.168.0.2:49890->/192.168.0.2:5701]“,tag2=in
    at com.sun.jmx.interceptor.DefaultMBeanServerInterceptor.getMBean(DefaultMBeanServerInterceptor.java:1095)
    at com.sun.jmx.interceptor.DefaultMBeanServerInterceptor.exclusiveUnregisterMBean(DefaultMBeanServerInterceptor.java:427)
    at com.sun.jmx.interceptor.DefaultMBeanServerInterceptor.unregisterMBean(DefaultMBeanServerInterceptor.java:415)
    at com.sun.jmx.mbeanserver.JmxMBeanServer.unregisterMBean(JmxMBeanServer.java:546)
    at com.hazelcast.jet.impl.metrics.jmx.JmxPublisher.whenComplete(JmxPublisher.java:107)
    ... 8 more
```

--- a/hazelcast-jet-core/src/main/java/com/hazelcast/jet/config/JobConfig.java
+++ b/hazelcast-jet-core/src/main/java/com/hazelcast/jet/config/JobConfig.java
@@ -116,8 +116,10 @@ public class JobConfig implements Serializable {
      * Sets whether Jet will scale the job up/down when a member is
      * added/removed from the cluster.
      * <p>
-     * If enabled and a member is added or removed, the job will automatically
-     * restart, see {@link Job#restart} for more details.
+     * If enabled and a member is added or removed, the job will
+     * automatically restart, see {@link Job#restart} for more details.
+     * In case of member addition, it will restart after a {@linkplain
+     * InstanceConfig#setScaleUpDelayMillis(long) delay}.
      * <p>
      * If disabled and a <em>member is added</em>, Jet takes no action and the
      * job will not use the added member; you have to manually {@linkplain
@@ -125,7 +127,7 @@ public class JobConfig implements Serializable {
      * shutdown or a failure), Jet suspends the job. You have to manually
      * {@linkplain Job#resume() resume} it.
      * <p>
-     * By default, auto-restart is enabled.
+     * By default, auto-scaling is enabled.
      *
      * @return {@code this} instance for fluent API
      */

--- a/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/metrics/jmx/JmxPublisher.java
+++ b/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/metrics/jmx/JmxPublisher.java
@@ -20,6 +20,8 @@ import com.hazelcast.internal.metrics.MetricsUtil;
 import com.hazelcast.jet.config.MetricsConfig;
 import com.hazelcast.jet.impl.metrics.MetricsPublisher;
 
+import javax.management.InstanceNotFoundException;
+import javax.management.MBeanRegistrationException;
 import javax.management.MBeanServer;
 import javax.management.MalformedObjectNameException;
 import javax.management.ObjectName;
@@ -51,6 +53,8 @@ public class JmxPublisher implements MetricsPublisher {
      */
     private final Map<ObjectName, MetricsMBean> mBeans = new HashMap<>();
     private final String domainPrefix;
+
+    private volatile boolean isShutdown;
 
     public JmxPublisher(String instanceName, String domainPrefix) {
         this.domainPrefix = domainPrefix;
@@ -87,7 +91,21 @@ public class JmxPublisher implements MetricsPublisher {
             }
             return bean;
         });
+        if (isShutdown) {
+            unregisterMBeanIgnoreError(metricData.objectName);
+        }
         mBean.setMetricValue(metricData.metric, metricData.unit, value);
+    }
+
+    private void unregisterMBeanIgnoreError(ObjectName objectName) {
+        try {
+            platformMBeanServer.unregisterMBean(objectName);
+        } catch (InstanceNotFoundException ignored) {
+            // Unregistration can by racy, we unregister from multiple places.
+            // See https://github.com/hazelcast/hazelcast-jet/issues/942
+        } catch (MBeanRegistrationException e) {
+            throw new RuntimeException(e);
+        }
     }
 
     @Override
@@ -103,11 +121,7 @@ public class JmxPublisher implements MetricsPublisher {
                 // remove entire bean if no metric is left
                 if (mBean.numAttributes() == 0) {
                     mBeans.remove(metricData.objectName);
-                    try {
-                        platformMBeanServer.unregisterMBean(metricData.objectName);
-                    } catch (Exception e) {
-                        throw new RuntimeException(e);
-                    }
+                    unregisterMBeanIgnoreError(metricData.objectName);
                 }
             } else {
                 metricData.wasPresent = false;
@@ -229,13 +243,16 @@ public class JmxPublisher implements MetricsPublisher {
 
     @Override
     public void shutdown() {
+        isShutdown = true;
+        ObjectName name;
         try {
-            ObjectName name = new ObjectName("com.hazelcast*" + ":instance=" + instanceNameEscaped + ",type=Metrics,*");
-            for (ObjectName bean : platformMBeanServer.queryNames(name, null)) {
-                platformMBeanServer.unregisterMBean(bean);
-            }
-        } catch (Exception e) {
+            name = new ObjectName("com.hazelcast*" + ":instance=" + instanceNameEscaped + ",type=Metrics,*");
+        } catch (MalformedObjectNameException e) {
             throw new RuntimeException("Exception when unregistering JMX beans", e);
+        }
+
+        for (ObjectName bean : platformMBeanServer.queryNames(name, null)) {
+            unregisterMBeanIgnoreError(bean);
         }
     }
 }

